src/vos/tests/vts_dtx.c

```diff
@@ -0,0 +1,1625 @@
+	uint64_t			 epoch;
+	uint64_t			 dkey_hash;
+	uint64_t			 saved_committable;
+	uint64_t			 saved_committed;
+	char				 dkey_buf[UPDATE_DKEY_SIZE];
+	char				 akey_buf[UPDATE_AKEY_SIZE];
+	char				 update_buf[UPDATE_BUF_SIZE];
+	int				 rc;
+
+	/* Assume I am the leader. */
+	vos_dtx_register_check_leader(vts_check_leader_true);
+
+	vts_dtx_prep_update(args, &xid, &val_iov, &dkey_iov, &dkey, dkey_buf,
+			    &akey, akey_buf, &iod, &sgl, &rex, update_buf,
+			    UPDATE_BUF_SIZE, UPDATE_REC_SIZE, &dkey_hash,
+			    &epoch, false);
+
+	rc = vts_dtx_begin(&xid, &args->oid, args->ctx.tc_co_hdl, epoch,
+			   dkey_hash, &conflict, NULL, 0, 1 /* init version */,
+			   DAOS_INTENT_UPDATE, &dth);
+	assert_int_equal(rc, 0);
+
+	rc = io_test_obj_update(args, epoch, &dkey, &iod, &sgl, dth, true);
+	assert_int_equal(rc, 0);
+
+	/* The DTX is 'prepared'. */
+	vts_dtx_end(dth);
+
+	/* Add former DTX into CoS cache. */
+	rc = vos_dtx_add_cos(args->ctx.tc_co_hdl, &args->oid, &xid,
+			     dkey_hash, epoch, false);
+	assert_int_equal(rc, 0);
+
+	vos_dtx_stat(args->ctx.tc_co_hdl, &stat);
+	saved_committable = stat.dtx_committable_count;
+	saved_committed = stat.dtx_committed_count;
+
+	/* Commit former DTX that will be removed from the CoS cache. */
+	rc = vos_dtx_commit(args->ctx.tc_co_hdl, &xid, 1);
+	assert_int_equal(rc, 0);
+
+	vos_dtx_stat(args->ctx.tc_co_hdl, &stat);
+	assert_true(saved_committable == stat.dtx_committable_count + 1);
+	assert_true(saved_committed == stat.dtx_committed_count - 1);
+}
+
+static void
+vts_dtx_commit_visibility(struct io_test_args *args, bool ext, bool punch_obj)
+{
+	struct dtx_handle		*dth = NULL;
+	struct dtx_id			 xid;
+	struct dtx_conflict_entry	 conflict = { 0 };
+	daos_iod_t			 iod = { 0 };
+	d_sg_list_t			 sgl = { 0 };
+	daos_recx_t			 rex = { 0 };
+	daos_key_t			 dkey;
+	daos_key_t			 akey;
+	d_iov_t				 val_iov;
+	d_iov_t				 dkey_iov;
+	uint64_t			 epoch;
+	uint64_t			 dkey_hash;
+	char				 dkey_buf[UPDATE_DKEY_SIZE];
+	char				 akey_buf[UPDATE_AKEY_SIZE];
+	char				 update_buf[UPDATE_BUF_SIZE];
+	char				 fetch_buf[UPDATE_BUF_SIZE];
+	int				 rc;
+
+	/* Assume I am the leader. */
+	vos_dtx_register_check_leader(vts_check_leader_true);
+
+	vts_dtx_prep_update(args, &xid, &val_iov, &dkey_iov, &dkey, dkey_buf,
+			    &akey, akey_buf, &iod, &sgl, &rex, update_buf,
+			    UPDATE_BUF_SIZE, UPDATE_REC_SIZE, &dkey_hash,
+			    &epoch, ext);
+
+	rc = vts_dtx_begin(&xid, &args->oid, args->ctx.tc_co_hdl, epoch,
+			   dkey_hash, &conflict, NULL, 0, 1 /* init version */,
+			   DAOS_INTENT_UPDATE, &dth);
+	assert_int_equal(rc, 0);
+
+	rc = io_test_obj_update(args, epoch, &dkey, &iod, &sgl, dth, true);
+	assert_int_equal(rc, 0);
+
+	/* The update DTX is 'prepared'. */
+	vts_dtx_end(dth);
+
+	memset(fetch_buf, 0, UPDATE_BUF_SIZE);
+	d_iov_set(&val_iov, fetch_buf, UPDATE_BUF_SIZE);
+	iod.iod_size = DAOS_REC_ANY;
+
+	rc = io_test_obj_fetch(args, epoch, &dkey, &iod, &sgl, true);
+	assert_int_equal(rc, 0);
```

Comment:
One note...when assertions fail, it is a pain to figure out what happened because they get printed in HEX.   It would be helpful to print non-zero rc in a more readable format (such as -1011 or d_err2str(rc)).   I've found this quite annoying when things go wrong in our existing tests.

Reply:
I agree with you. I would suggest to make another patch to replace assert_int_equal() for the whole vos_test.